tests/lms/services/exceptions_test.py

```diff
@@ -54,8 +62,143 @@ def test_when_theres_a_response_it_uses_it_in_str(
+            "response": {"status": expected_status, "body": '{"foo": "bar"}'},
+        }
+
+    @pytest.mark.parametrize(
+        "cause,expected_exception_string",
+        [
+            (requests.RequestException(), "RequestException()"),
+            (requests.ConnectionError(), "ConnectionError()"),
+            (requests.TooManyRedirects(), "TooManyRedirects()"),
+            (requests.ConnectTimeout(), "ConnectTimeout()"),
+            (requests.ReadTimeout(), "ReadTimeout()"),
+        ],
+    )
+    def test_it_raises_CanvasAPIServerError_for_all_other_requests_errors(
+        self, cause, expected_exception_string
+    ):
+        with pytest.raises(
+            CanvasAPIServerError, match="Calling the Canvas API failed"
+        ) as exc_info:
+            CanvasAPIError.raise_from(cause)
+
+        raised_exception = exc_info.value
+        assert raised_exception.__cause__ == cause
+        # For these kinds of errors no response (either successful or
+        # unsuccessful) was ever received from Canvas (for example: the network
+        # request timed out) so there's nothing to set as the response
+        # property.
+        assert raised_exception.response is None
+        assert raised_exception.details == {
+            "exception": expected_exception_string,
+            "response": None,
+            "validation_errors": None,
+        }
+
+    def test_it_raises_CanvasAPIServerError_for_a_successful_but_invalid_response(
+        self, canvas_api_invalid_response
+    ):
+        cause = ValidationError("The response was invalid.")
+        cause.response = canvas_api_invalid_response
+
+        with pytest.raises(
+            CanvasAPIServerError, match="Calling the Canvas API failed"
+        ) as exc_info:
+            CanvasAPIError.raise_from(cause)
+
+        raised_exception = exc_info.value
+        assert raised_exception.__cause__ == cause
+        assert raised_exception.response == canvas_api_invalid_response
+        assert raised_exception.details == {
+            "exception": "Unable to process the contained instructions",
+            "response": {"body": "Invalid", "status": "200 OK"},
+            "validation_errors": "The response was invalid.",
+        }
+
+    @pytest.fixture
+    def canvas_api_invalid_response(self):
+        """Return a successful (200 OK) response with an invalid body."""
+        httpretty.register_uri(
+            httpretty.GET, "https://example.com", priority=1, status=200, body="Invalid"
+        )
+        return requests.get("https://example.com")
+
+    @staticmethod
+    def _requests_exception(**kwargs):
+        httpretty.register_uri(
+            httpretty.GET,
+            "https://example.com",
+            priority=1,
+            body=json.dumps({"foo": "bar"}),
+            **kwargs
+        )
+
+        response = requests.get("https://example.com")
+
+        try:
+            response.raise_for_status()
+        except requests.RequestException as err:
```

Comment:
What are the limitations of constructing the `RequestException` manually vs dynamically generating them as you are here?

Reply:
:man_shrugging: I _think_ you could construct a `requests.RequestException` with all the necessary fields manually. `requests` does have some fields that it refuses to let you set, so we might run into that (but you can get around that by mocking). But I think it's just nicer to use the real `requests` to create the exception (using `httpretty` to trick `requests` into doing that) because then we're testing in better integration with `request`.